File: MyProject/Source/MyProject/Private/Game/Libs/DelayHandle/DelayHandleMgrBase.cpp
#include "MyProject.h"
#include "DelayHandleMgrBase.h"
#include "DelayHandleObject.h"
#include "DelayAddParam.h"
#include "DelayDelParam.h"
#include "IDelayHandleItem.h"
#include "UtilContainers.h"

// 只有没有添加到列表中的才能添加
bool DelayHandleMgrBase::existAddList(IDelayHandleItem* delayObject)
{
	for(auto item : m_deferredAddQueue.getList())
	{
		if (item->m_delayObject == delayObject)
		{
			return true;
		}
	}

	return false;
}

// 只有没有添加到列表中的才能添加
bool DelayHandleMgrBase::existDelList(IDelayHandleItem* delayObject)
{
	for(auto item : m_deferredDelQueue.getList())
	{
		if (item->m_delayObject == delayObject)
		{
			return true;
		}
	}

	return false;
}

// 从延迟添加列表删除一个 Item
void DelayHandleMgrBase::delFromDelayAddList(IDelayHandleItem* delayObject)
{
	for(auto item : m_deferredAddQueue.getList())
	{
		if (item->m_delayObject == delayObject)
		{
			UtilVector::Remove(m_deferredAddQueue.getList(), item);
		}
	}
}

// 从延迟删除列表删除一个 Item
void DelayHandleMgrBase::delFromDelayDelList(IDelayHandleItem* delayObject)
{
	for(auto item : m_deferredDelQueue.getList())
	{
		if (item->m_delayObject == delayObject)
		{
			UtilVector::Remove(m_deferredDelQueue.getList(), item);
		}
	}
}

void DelayHandleMgrBase::processDelayObjects()
{
	if (0 == m_loopDepth)       // 只有全部退出循环后，才能处理添加删除
	{
		if (m_deferredAddQueue.Count() > 0)
		{
			for (int idx = 0; idx < m_deferredAddQueue.Count(); idx++)
			{
				addObject(m_deferredAddQueue[idx]->m_delayObject, ((DelayAddParam*)(m_deferredAddQueue[idx]->m_delayParam))->m_priority);
			}

			m_deferredAddQueue.Clear();
		}

		if (m_deferredDelQueue.Count() > 0)
		{
			for (int idx = 0; idx < m_deferredDelQueue.Count(); idx++)
			{
				delObject(m_deferredDelQueue[idx]->m_delayObject);
			}

			m_deferredDelQueue.Clear();
		}
	}
}

DelayHandleMgrBase::DelayHandleMgrBase()
{
	m_loopDepth = 0;
}

void DelayHandleMgrBase::addObject(IDelayHandleItem* delayObject, float priority)
{
	if (m_loopDepth > 0)
	{
		if (!existAddList(delayObject))        // 如果添加列表中没有
		{
			if (existDelList(delayObject))    // 如果已经添加到删除列表中
			{
				delFromDelayDelList(delayObject);
			}

			DelayHandleObject* delayHandleObject = new DelayHandleObject();
			delayHandleObject->m_delayParam = new DelayAddParam();
			m_deferredAddQueue.Add(delayHandleObject);

			delayHandleObject->m_delayObject = delayObject;
			((DelayAddParam*)(delayHandleObject->m_delayParam))->m_priority = priority;
		}
	}
}

void DelayHandleMgrBase::delObject(IDelayHandleItem* delayObject)
{
	if (m_loopDepth > 0)
	{
		if (!existDelList(delayObject))
		{
			if (existAddList(delayObject))    // 如果已经添加到删除列表中
			{
				delFromDelayAddList(delayObject);
			}

			delayObject->setClientDispose();

			DelayHandleObject* delayHandleObject = new DelayHandleObject();
			delayHandleObject->m_delayParam = new DelayDelParam();
			m_deferredDelQueue.Add(delayHandleObject);
			delayHandleObject->m_delayObject = delayObject;
		}
	}
}

void DelayHandleMgrBase::incDepth()
{
	++m_loopDepth;
}

void DelayHandleMgrBase::decDepth()
{
	--m_loopDepth;
	processDelayObjects();
}

bool DelayHandleMgrBase::bInDepth()
{
	return m_loopDepth > 0;
}

File: MyProject/Content/MyAsset/Lua/MyLua/Libs/DelayHandle/DelayHandleMgrBase.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");
MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DelayHandle.DelayHandleObject");
MLoader("MyLua.Libs.DelayHandle.DelayAddParam");
MLoader("MyLua.Libs.DelayHandle.DelayDelParam");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "DelayHandleMgrBase";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mDeferredAddQueue = GlobalNS.new(GlobalNS.MList);
    self.mDeferredDelQueue = GlobalNS.new(GlobalNS.MList);

    self.mLoopDepth = 0;
end

function M:dtor()

end

function M:addObject(delayObject, priority)
    if (self.mLoopDepth > 0) then
        if (not self:existAddList(delayObject)) then       -- 如果添加列表中没有
            if (self:existDelList(delayObject)) then   -- 如果已经添加到删除列表中
                self:delFromDelayDelList(delayObject);
            end

            local delayHandleObject = GlobalNS.new(GlobalNS.DelayHandleObject);
            delayHandleObject.mDelayParam = GlobalNS.new(GlobalNS.DelayAddParam);
            self.mDeferredAddQueue:Add(delayHandleObject);

            delayHandleObject.mDelayObject = delayObject;
            delayHandleObject.mDelayParam.mPriority = priority;
        end
    end
end

function M:removeObject(delayObject)
    if (self.mLoopDepth > 0) then
        if (not self:existDelList(delayObject)) then
            if (self:existAddList(delayObject)) then    -- 如果已经添加到删除列表中
                self:delFromDelayAddList(delayObject);
            end

            delayObject:setClientDispose();

            local delayHandleObject = GlobalNS.new(GlobalNS.DelayHandleObject);
            delayHandleObject.mDelayParam = GlobalNS.new(GlobalNS.DelayDelParam);
            self.mDeferredDelQueue:Add(delayHandleObject);
            delayHandleObject.mDelayObject = delayObject;
        end
    end
end

-- 只有没有添加到列表中的才能添加
function M:existAddList(delayObject)
    for _, item in ipairs(self.mDeferredAddQueue:list()) do
        if (item.mDelayObject == delayObject) then
            return true;
        end
    end

    return false;
end

-- 只有没有添加到列表中的才能添加
function M:existDelList(delayObject)
    for _, item in ipairs(self.mDeferredAddQueue:list()) do
        if (item.mDelayObject == delayObject) then
            return true;
        end
    end

    return false;
end

-- 从延迟添加列表删除一个 Item
function M:delFromDelayAddList(delayObject)
    for _, item in ipairs(self.mDeferredAddQueue:list()) do
        if (item.mDelayObject == delayObject) then
            self.mDeferredAddQueue:Remove(item);
        end
    end
end

-- 从延迟删除列表删除一个 Item
function M:delFromDelayDelList(delayObject)
    for _, item in ipairs(self.mDeferredDelQueue:list()) do
        if (item.mDelayObject == delayObject) then
            self.mDeferredDelQueue:Remove(item);
        end
    end
end

function M:processDelayObjects()
    if 0 == self.mLoopDepth then       -- 只有全部退出循环后，才能处理添加删除
        if (self.mDeferredAddQueue:Count() > 0) then
            local idx = 0;
            for idx = 0, self.mDeferredAddQueue:Count() - 1, 1 do
                self:addObject(self.mDeferredAddQueue:at(idx).mDelayObject, self.mDeferredAddQueue:at(idx).mDelayParam.mPriority);
            end

            self.mDeferredAddQueue:Clear();
        end

        if (self.mDeferredDelQueue:Count() > 0) then
            local idx = 0;
            for idx = 0, self.mDeferredDelQueue:Count() - 1, 1 do
                self:removeObject(self.mDeferredDelQueue:at(idx).mDelayObject);
            end

            self.mDeferredDelQueue:Clear();
        end
    end
end

function M:incDepth()
    self.mLoopDepth = self.mLoopDepth + 1;
end

function M:decDepth()
    self.mLoopDepth = self.mLoopDepth - 1;
    self:processDelayObjects();
end

function M:bInDepth()
    return self.mLoopDepth > 0;
end

return M;